runtime/jcl/common/criusupport.c

```diff
@@ -44,3 +54,94 @@ Java_com_ibm_jvm_CRIUSupport_isCRIUSupportEnabledImpl(JNIEnv *env, jclass unused
 
 	return res;
 }
+
+jint JNICALL
+Java_com_ibm_jvm_CRIUSupport_checkpointJVMImpl(JNIEnv *env,
+							jclass unused,
+							jobject checkPointDir,
+							jboolean keepRunning,
+							jboolean shellJob,
+							jboolean extUnixSupport,
+							jint logLevel,
+							jobject logFile)
+{
+	jint rc = UNSUPPORTED_OPERATION;
+	J9VMThread *currentThread = (J9VMThread*)env;
+	J9JavaVM *vm = currentThread->javaVM;
+	J9InternalVMFunctions *vmFuncs = vm->internalVMFunctions;
+	char directoryBuf[STRING_BUFFER_SIZE] = {0};
+	char *directoryUTF = NULL;
+	char logFileBuf[STRING_BUFFER_SIZE] = {0};
+	char *logFileUTF = NULL;
+
+	if (J9_ARE_ALL_BITS_SET(vm->extendedRuntimeFlags2, J9_EXTENDED_RUNTIME2_ENABLE_CRIU_SUPPORT)) {
+#if defined(LINUX)
+		j9object_t cpDir = NULL;
+		j9object_t log = NULL;
+		IDATA dirFD = 0;
+		PORT_ACCESS_FROM_VMC(currentThread);
+
+		vmFuncs->internalEnterVMFromJNI(currentThread);
+
+		cpDir = J9_JNI_UNWRAP_REFERENCE(checkPointDir);
+		log = J9_JNI_UNWRAP_REFERENCE(logFile);
+
+		directoryUTF = vmFuncs->copyStringToUTF8WithMemAlloc(currentThread, cpDir, J9_STR_NULL_TERMINATE_RESULT, "", 0, directoryBuf, STRING_BUFFER_SIZE, NULL);
+		if (NULL == directoryUTF) {
+			vmFuncs->setNativeOutOfMemoryError(currentThread, 0, 0);
+			goto release;
+		}
+
+		logFileUTF = vmFuncs->copyStringToUTF8WithMemAlloc(currentThread, log, J9_STR_NULL_TERMINATE_RESULT, "", 0, logFileBuf, STRING_BUFFER_SIZE, NULL);
+		if (NULL == logFileUTF) {
+			vmFuncs->setNativeOutOfMemoryError(currentThread, 0, 0);
+			goto freeDir;
+		}
+
+		dirFD = open(directoryUTF, O_DIRECTORY);
+		if (dirFD < 0) {
+			rc = INVALID_ARGUEMENTS;
+			goto freeLog;
+		}
+
+		criu_set_images_dir_fd(dirFD);
+		criu_set_shell_job(shellJob);
+		criu_set_log_level(logLevel);
+		criu_set_log_file(logFileUTF);
+		criu_set_leave_running(keepRunning);
+		criu_set_ext_unix_sk(extUnixSupport);
+
+		if (FALSE == vmFuncs->jvmCheckpointHooks(currentThread)) {
+			rc = JVM_CHECKPOINT_FAILURE;
+			goto freeLog;
+		}
+
+		if (0 > criu_dump()) {
```

Comment:
Do you want to take exclusive vmaccess around the criu_dump() call?

Reply:
Since we don't have any hooks yet, it is not necessary. But we can add it now.